src/app/source-file-tabs/source-file-tabs.component.ts

```diff
@@ -0,0 +1,121 @@
+  styleUrls: ['./source-file-tabs.component.scss']
+})
+export class SourceFileTabsComponent implements OnInit {
+  @Input() workflowId: number;
+  @Input() descriptorType: ToolDescriptor.TypeEnum;
+  @Input() set version(value: WorkflowVersion) {
+    if (value != null) {
+      this.currentVersion = value;
+      this.setupVersionFileTabs();
+    }
+  }
+  loading = true;
+  currentVersion: WorkflowVersion;
+  files: SourceFile[];
+  filteredFiles: SourceFile[];
+  currentFile: SourceFile;
+  fileTypes: SourceFile.TypeEnum[];
+  currentFileType: SourceFile.TypeEnum;
+  validationMessage: Object;
+  customDownloadHREF: SafeUrl;
+  customDownloadPath: String;
+  filePath: String;
+
+  constructor(
+    private workflowsService: WorkflowsService,
+    private fileService: FileService,
+    private descriptorTypeCompatService: DescriptorTypeCompatService
+  ) {}
+
+  ngOnInit() {
+    this.setupVersionFileTabs();
+  }
+
+  setupVersionFileTabs() {
+    this.loading = true;
+    this.workflowsService
+      .getWorkflowVersionsSourcefiles(this.workflowId, this.currentVersion.id)
+      .pipe(
+        finalize(() => {
+          this.loading = false;
+        })
+      )
+      .subscribe(
+        (sourceFiles: SourceFile[]) => {
+          this.files = sourceFiles;
+          this.fileTypes = [];
+          this.files.forEach((file: SourceFile) => {
+            if (!this.fileTypes.includes(file.type)) {
+              this.fileTypes.push(file.type);
+            }
+          });
+          if (this.fileTypes.length > 0) {
+            this.changeFileType(this.fileTypes[0]);
+          }
+        },
+        error => {
+          console.log(error);
+        }
+      );
+  }
+
+  /**
+   * Sets the file type to display and loads validation messages
+   * @param fileType
+   */
+  changeFileType(fileType: SourceFile.TypeEnum) {
+    this.currentFileType = fileType;
+
+    this.filteredFiles = this.files.filter((file: SourceFile) => {
+      return file.type === this.currentFileType;
+    });
+
+    if (this.filteredFiles.length > 0) {
+      this.selectFile(this.filteredFiles[0]);
+    }
+    this.currentVersion.validations.forEach((validation: Validation) => {
+      this.validationMessage = null;
+      if (validation.type === this.currentFileType && !validation.valid) {
+        this.validationMessage = JSON.parse(validation.message);
+      }
+    });
+  }
+
+  selectFile(file: SourceFile) {
+    this.currentFile = file;
+    this.customDownloadHREF = this.fileService.getFileData(this.currentFile.content);
+    this.customDownloadPath = this.fileService.getFileName(this.currentFile.path);
+    this.filePath = this.getDescriptorPath();
+  }
+
+  matTabChange(event: MatTabChangeEvent) {
+    const fileType: SourceFile.TypeEnum = this.fileTypes[event.index];
+    this.changeFileType(fileType);
+  }
+
+  matSelectChange(event: MatSelectChange) {
+    this.selectFile(event.value);
+  }
+
+  getDescriptorPath(): string {
```

Comment:
convert to static function if possible (which will making moving stuff to a service easier)